In [69]:
url = 'https://www.basketball-reference.com/leagues/NBA_2010_per_game.html'

In [70]:
# Import pandas
import pandas as pd

In [71]:
#Use a for loop to  collect data from 2010 à 2024
df = pd.DataFrame()

for year in range(2010,2025) :
    url = f'https://www.basketball-reference.com/leagues/NBA_{year}_per_game.html'
    data = pd.read_html(url, header = 0)[0]
    df = pd.concat([df,data], axis = 0).reset_index(drop = True)

In [72]:
# Display df
df.head()

Rk           Player   Age Team Pos     G    GS    MP    FG   FGA  ...  \
0  1.0     Kevin Durant  21.0  OKC  SF  82.0  82.0  39.5   9.7  20.3  ...   
1  2.0     LeBron James  25.0  CLE  SF  76.0  76.0  39.0  10.1  20.1  ...   
2  3.0  Carmelo Anthony  25.0  DEN  SF  69.0  69.0  38.2  10.0  21.8  ...   
3  4.0      Kobe Bryant  31.0  LAL  SG  73.0  73.0  38.8   9.8  21.5  ...   
4  5.0      Dwyane Wade  28.0  MIA  SG  77.0  77.0  36.3   9.3  19.6  ...   

   ORB  DRB  TRB  AST  STL  BLK  TOV   PF   PTS                 Awards  
0  1.3  6.3  7.6  2.8  1.4  1.0  3.3  2.1  30.1          MVP-2,AS,NBA1  
1  0.9  6.4  7.3  8.6  1.6  1.0  3.4  1.6  29.7   MVP-1,DPOY-4,AS,NBA1  
2  2.2  4.4  6.6  3.2  1.3  0.4  3.0  3.3  28.2          MVP-6,AS,NBA2  
3  1.1  4.3  5.4  5.0  1.5  0.3  3.2  2.6  27.0  MVP-3,DPOY-12,AS,NBA1  
4  1.4  3.5  4.8  6.5  1.8  1.1  3.3  2.4  26.6  MVP-5,DPOY-10,AS,NBA1  

[5 rows x 31 columns]

In [73]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9731 entries, 0 to 9730
Data columns (total 31 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Rk      9716 non-null   float64
 1   Player  9731 non-null   object 
 2   Age     9716 non-null   float64
 3   Team    9716 non-null   object 
 4   Pos     9716 non-null   object 
 5   G       9716 non-null   float64
 6   GS      9716 non-null   float64
 7   MP      9716 non-null   float64
 8   FG      9716 non-null   float64
 9   FGA     9716 non-null   float64
 10  FG%     9667 non-null   float64
 11  3P      9716 non-null   float64
 12  3PA     9716 non-null   float64
 13  3P%     8730 non-null   float64
 14  2P      9716 non-null   float64
 15  2PA     9716 non-null   float64
 16  2P%     9599 non-null   float64
 17  eFG%    9667 non-null   float64
 18  FT      9716 non-null   float64
 19  FTA     9716 non-null   float64
 20  FT%     9173 non-null   float64
 21  ORB     9716 non-null   float64
 22  

In [75]:
# We want to  know the names of the columns of our data base.
df.columns

Index(['Rk', 'Player', 'Age', 'Team', 'Pos', 'G', 'GS', 'MP', 'FG', 'FGA',
       'FG%', '3P', '3PA', '3P%', '2P', '2PA', '2P%', 'eFG%', 'FT', 'FTA',
       'FT%', 'ORB', 'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS',
       'Awards'],
      dtype='object')

# **It will be good to understand what these names means. So that we can easily continuous Our analyse**

### Signification of the variables

1. **Rk** : the player's rank or position in a certain category or ranking.
2. **Age** : Age of the player.
3. **Team** : name of the player's team.
4. **Pos** : the player's position on the court.
5. **G** : number of games the player has participated in.
6. **GS** : number of games the player started in the starting lineup.
7. **MP** : total number of minutes the player has played.
8. **FG** : number of field goals made.
9. **FGA** : number of field goals attempted.
10. **FG%** : percentage of successful field goals.
11. **3P** : number of 3-point shots made.
12. **3PA** : number of 3-point shots attempted.
13. **3P%** : percentage of successful 3-point shots.
14. **2P** : number of 2-point shots made.
15. **2PA** : number of 2-point shots attempted.
16. **2P%** : percentage of successful 2-point shots.
17. **eFG%** : an advanced statistic that adjusts the field goal percentage.
18. **FT** : Number of free throws made.
19. **FTA** : number of free throws attempted.
20. **FT%** : percentage of successful free throws.
21. **ORB** : number of offensive rebounds grabbed by the player.
22. **DRB** : number of defensive rebounds grabbed by the player.
23. **TRB** : total number of rebounds, the sum of offensive and defensive rebounds.
24. **AST** : number of assists (passes that directly lead to a basket).
25. **STL** : number of successful steals by the player.
26. **BLK** : number of successful blocks by the player.
27. **TOV** : number of times the player has lost possession of the ball.
28. **PF** : number of personal fouls committed by the player.
29. **PTS** : total number of points scored by the player.
30. **Awards** : list of awards or distinctions received by the player.


In [76]:
# Dimension of df
df.shape

(9731, 31)

In [77]:
# Check data types
df.dtypes

Rk        float64
Player     object
Age       float64
Team       object
Pos        object
G         float64
GS        float64
MP        float64
FG        float64
FGA       float64
FG%       float64
3P        float64
3PA       float64
3P%       float64
2P        float64
2PA       float64
2P%       float64
eFG%      float64
FT        float64
FTA       float64
FT%       float64
ORB       float64
DRB       float64
TRB       float64
AST       float64
STL       float64
BLK       float64
TOV       float64
PF        float64
PTS       float64
Awards     object
dtype: object

In [78]:
# Check for null values
df.isnull().any().sum()

30

In [79]:
# Check for null values.
missing_value = df.isnull().sum()

missing_value

Rk          15
Player       0
Age         15
Team        15
Pos         15
G           15
GS          15
MP          15
FG          15
FGA         15
FG%         64
3P          15
3PA         15
3P%       1001
2P          15
2PA         15
2P%        132
eFG%        64
FT          15
FTA         15
FT%        558
ORB         15
DRB         15
TRB         15
AST         15
STL         15
BLK         15
TOV         15
PF          15
PTS         15
Awards    8851
dtype: int64

We can remark that some variables as Team, Awards are qualitatives variables. So we will separate our variables in two categories : float and object variables

In [80]:
float_variables = df.select_dtypes(include=['float']).columns.tolist()
object_variables = df.select_dtypes(include=['object']).columns.tolist()

print("float variables:", float_variables)
print("object variables:", object_variables)


float variables: ['Rk', 'Age', 'G', 'GS', 'MP', 'FG', 'FGA', 'FG%', '3P', '3PA', '3P%', '2P', '2PA', '2P%', 'eFG%', 'FT', 'FTA', 'FT%', 'ORB', 'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS']
object variables: ['Player', 'Team', 'Pos', 'Awards']


**Let's treate numerical variables.**

> We remark that the variable 3P% (percentage of successful 3-point shots) has 1001 missing value. 10% of our data. We will not imput the missing value. We will drop this variable at the end of our data preprocessing.

* Another variables if Awards (Categorial variable) who has 8851 missing value. That is too much. We will also drop it.

* In addition the variable Rk (the player's rank or position in a certain category or ranking), is not a variable that we can imput because of the missing value. Unformtunetly, we will keep it like that without modification.

In [ ]:
# We will now go trough each variables and look for the descriptive statistic and now decide if we will input by the mean or the median.

> Age

In [81]:
df.Age.describe()

count    9716.000000
mean       26.314121
std         4.171557
min        19.000000
25%        23.000000
50%        26.000000
75%        29.000000
max        43.000000
Name: Age, dtype: float64

The mean and the median look like the same. We can imput missing value by mean

In [44]:
# Now we will do the same thing but by using a loop. On our numerical variables.

In [82]:
variable_input = ['G', 'GS', 'MP', 'FG', 'FGA', 'FG%', '3P', '3PA', '2P', '2PA', '2P%', 'eFG%', 'FT', 'FTA', 'FT%', 'ORB', 'DRB', 'TRB', 'AST', 'STL', 'BLK', 'TOV', 'PF', 'PTS']

for var in variable_input :
    print(df[var].describe())

count    9716.000000
mean       43.509057
std        25.667309
min         1.000000
25%        20.000000
50%        46.000000
75%        67.000000
max        84.000000
Name: G, dtype: float64
count    9716.000000
mean       20.075957
std        25.954269
min         0.000000
25%         0.000000
50%         6.000000
75%        35.000000
max        83.000000
Name: GS, dtype: float64
count    9716.000000
mean       19.261311
std         9.432216
min         0.000000
25%        11.700000
50%        18.800000
75%        27.000000
max        43.500000
Name: MP, dtype: float64
count    9716.000000
mean        2.990212
std         2.147003
min         0.000000
25%         1.300000
50%         2.500000
75%         4.200000
max        11.500000
Name: FG, dtype: float64
count    9716.000000
mean        6.641951
std         4.498497
min         0.000000
25%         3.200000
50%         5.600000
75%         9.100000
max        24.500000
Name: FGA, dtype: float64
count    9667.000000
mean        0.

After this process, we selected the following variables to apply the mean for the missing values.

We selected these variables because the mean is approximately equal to the median.

The variables are:

* ['Age', 'G', 'MP', 'FGA', 'FG%', '2P%', 'eFG%', 'FT', 'BLK', 'PF']

In [83]:
mean_fillna = ['Age', 'G', 'MP', 'FGA', 'FG%', '2P%', 'eFG%', 'FT', 'BLK', 'PF']

for var in mean_fillna:
    df[var].fillna(df[var].mean(), inplace=True)

# This code work but, Python just tell to us that in pandas 3.0, this will change.

<ipython-input-83-100a3a7f9c3b>:4: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df[var].fillna(df[var].mean(), inplace=True)


Now, for variables where the mean is larger than the median, or much larger than the median value, we choose the median for the input.

> These are:

* ['GS','FG','3P','3PA','2P','2PA','FT', 'FTA', 'FT%', 'ORB', 'DRB', 'TRB', 'AST', 'STL','TOV','PTS']

In [91]:
median_fillna = ['GS','FG','3P','3PA','2P','2PA','FT', 'FTA', 'FT%', 'ORB', 'DRB', 'TRB', 'AST', 'STL','TOV','PTS']

for var in median_fillna:
    df[var] = df[var].fillna(df[var].median())

In [53]:
# Let's treate numerical categorial value.

We will replace the missing value by the mode.

In [67]:
mode_fillna = ['Team','Pos']

for var in mode_fillna:
    df['var'] = df[var].fillna(df[var].mode())

Now we will drop the column Awards and 3P%.

In [85]:
df.drop(["Awards", "3P%"], axis=1, inplace=True)

In [92]:
df.isna().sum()

Rk        15
Player     0
Age        0
Team      15
Pos       15
G          0
GS         0
MP         0
FG         0
FGA        0
FG%        0
3P         0
3PA        0
2P         0
2PA        0
2P%        0
eFG%       0
FT         0
FTA        0
FT%        0
ORB        0
DRB        0
TRB        0
AST        0
STL        0
BLK        0
TOV        0
PF         0
PTS        0
var        0
dtype: int64

We export now the data to excel file.

In [94]:
df.to_excel('Data_clean.xlsx', header = True)